Add max concurrent requests configuration option to coordinator

psarna opened the pull request:
This series approaches issue #7072 and provides a very simple mechanism for limiting the number of concurrent CQL requests being served on a shard. Once the limit is hit, new requests will be instantly refused and OverloadedException will be returned to the client.
This mechanism has many improvement opportunities:
 * shedding requests gradually instead of having one hard limit,
 * having more than one limit per different types of queries (reads, writes, schema changes, ...),
 * not using a preconfigured value at all, and instead figuring out the limit dynamically,
 * etc.

... and none of these are taken into account in this series, which only adds a very basic configuration variable. The variable can be updated live without a restart - it can be done by updating the .yaml file and triggering a configuration re-read via sending the SIGHUP signal to Scylla.

The default value for this parameter is a very large number, which translates to effectively not shedding any requests at all.

Refs #7072

avikivity commented on transport/server.cc:
```diff
@@ -618,6 +618,11 @@ future<> cql_server::connection::process_request() {
                     f.length, mem_estimate, _server._max_request_size));
         }
 
+        if (_server._config.max_concurrent_requests && _server._requests_serving > _server._config.max_concurrent_requests) {
```
Why not default it to a very large value? That saves an extra check.

I'm not a fan of using special values to indicate meta-information (in this case, "ignore this value").

vladzcloudius replied:
something like `std::numeric_limits<typeof(_server._requests_serving)>::max()`

psarna replied:
done

avikivity commented on transport/controller.cc:
```diff
@@ -70,6 +71,14 @@ future<> controller::do_start_server() {
+            observers.emplace_back(make_lw_shared<utils::observer<uint64_t>>(cfg.max_concurrent_cql_requests.observe([cserver] (size_t value) {
+                logger.info("Setting max concurrent CQL requests threshold to {}", value);
+                cserver->local().set_max_concurrent_requests(value);
+            })));
```
You can also use updateable_value. It is designed to be move-constructible os you can just place it in config. It is also cross-shard-safe.

psarna replied:
done

psarna commented:
v2:
 * switched from tedious manual `observe` handling to simply using `updateable_value`
 * dropped special meaning of `0` in favor of using a very large number as a default number of allowed concurrent requests

vladzcloudius commented on transport/server.cc:
```diff
@@ -618,6 +618,11 @@ future<> cql_server::connection::process_request() {
 
+        if (_server._requests_serving > _server._config.max_concurrent_requests) {
+            return make_exception_future<>(
+                    exceptions::overloaded_exception(format("too many in-flight requests: {}", _server._requests_serving)));
```
a nitpick: maybe don't break that into two lines.

denesb replied:
Won't this always be the same number -- the configured limit? If this error is user facing, maybe worth mentioning how this limit was determined, i.e. mention the configuration item name.

psarna replied:
Right now it is, but the idea is to eventually™ make the threshold more elastic and automatically adapted. Then, the message will still make sense and potentially mention different values.

denesb replied:
Right, makes sense.

denesb replied:
It might still be worth mentioning how the limit was configured, it might prevent unnecessary questions from users. Something like https://github.com/scylladb/scylla/blob/1adf2cc848694ce9d2558aaea60800887d64057a/flat_mutation_reader.cc#L132-L135

This need not block this patchset, can be done as a follow-up -- if you agree to it.

psarna replied:
Sure, good idea. By the way, it's definitely not a blocker for this patchset, since it was queued 3 hours ago (:

denesb replied:
Oh, didn't notice. :)

vladzcloudius commented on transport/server.hh:
```diff
@@ -104,6 +104,7 @@ struct cql_query_state {
 struct cql_server_config {
     ::timeout_config timeout_config;
     size_t max_request_size;
+    size_t max_concurrent_requests;
```
You use different types for `max_concurrent_requests` (size_t) and for the corresponding cfg parameter `named_value<uint64_t> max_concurrent_cql_requests` .

While on x86_64 it's likely going to be the same `unsigned long long` however it would be safer to use the exactly same integer type (uint64_t) for both.

psarna replied:
ouch, rookie oversight on my part, will fix in a minute

avikivity replied:
uint64_t seems excessive. Why would we want > 4 billion concurrent requests on the same shard?

psarna commented:
v3:
 * unified the parameter type to uint64_t, it was mistakenly size_t in a couple of places

vladzcloudius reviewed (approved):
LGTM

avikivity commented on db/config.cc:
```diff
@@ -745,6 +745,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , schema_registry_grace_period(this, "schema_registry_grace_period", value_status::Used, 1,
         "Time period in seconds after which unused schema versions will be evicted from the local schema registry cache. Default is 1 second.")
+    , max_concurrent_cql_requests(this, "max_concurrent_cql_requests", value_status::Used, std::numeric_limits<uint64_t>::max(),
+        "Maximum number of concurrent CQL requests a single shard can handle before it starts shedding extra load. Default 0 means that no requests will be shed.")
```
0 is outdated (I presume)
uint64_t is excessive.

Perhaps the name should emphasize the per-shardness. It's easy to overlook.

psarna replied:
It is outdated. I'll fix the description, the name, and switch to u32 tomorrow.

vladzcloudius replied:
@psarna @avikivity it has to have the same type as `_requests_serving` if we want to keep the `limits` trick. 
If we change `max_concurrent_cql_requests` to u32 - we need to change the ` _requests_serving` type too (and I guess we don't want to do that) or reconsider the default value the relevant logic. 

I think keeping `max_concurrent_cql_requests` as u64 despite that probably being excessive is an acceptable compromise.

psarna replied:
Well, it doesn't really *have to* be the same type as long as we agree that it's not realistic of  `_requests_serving` to ever grow beyond 4294967295. But on the other hand I agree that it makes more sense -- e.g. if a bug in `_requests_serving` causes it to grow indefinitely, `max_concurrent_cql_requests` would start acting really weird as well. I'm fine with rolling it back to `uint64_t`, but since @avikivity  requested the change I'll wait for his comment as well.

avikivity replied:
I don't see a problem with u32. We can change _requests_serving to that type too if it makes anyone more comfortable, but it isn't really necessary.

vladzcloudius replied:
If we compare two integer types of different size the compiler will have to extend the smaller one anyway during the comparison. 

Also, since the parameter is not a naked uintX but a named_value<uintX> and the value itself is located after two string_view values and currently string_view has two following non-static fields:

```
const char* ptr_;
  size_type length_;
```

And unless I'm missing something...
No matter what integer type we are going to use it's going to be 64-bit aligned (since we do not use packing for named_value).

Therefore if we use uint64 or uint32 we are going to use/waste the same amount of memory.
Meaning using uint32 is not going to save any.

And then back to what @psarna wrote and what I meant all alone - using the same type for both ` _requests_serving` and `max_concurrent_cql_requests` while not a must is safer given the logic we intend to use.

vladzcloudius replied:
Oh, I noticed that I did missed an uint8_t value after the T in the named_value. 
So, using uint32 here would actually do save some space here. 

However, I still think that using uint64_t would be safer have.
And since we decided not to (as it seems) then we can no longer use the `limit::max()` trick here.

avikivity commented on transport/server.hh:
```diff
@@ -104,6 +104,7 @@ struct cql_query_state {
 struct cql_server_config {
     ::timeout_config timeout_config;
     size_t max_request_size;
+    uint64_t max_concurrent_requests;
```
Shouldn't this be updateable_value?

avikivity replied:
Ah, I see a later patch changes it.

psarna commented:
v4:
 * switched from using uint64_t to sufficient uint32_t as parameter type
 * dropped mentioning `0` as a special value of a param, which is no longer the case
 * added `_per_shard` to the name to make it explicit that each shard has its own threshold for requests

avikivity commented:
I hate to bring this up in v4, but I realized now we have redis, alternator, and thrift actiing as coordinators.

We have two ways to proceed:
 - decide that redis and thrift need individual limiters
 - decide that we need a global coordinator limit

In the first case, we need do nothing - we don't have implement redis/thrift limiters now.
In the second case, we need to at least rename the variable so it is protocol neutral (we don't have to apply the limits to redis/thrift).

Since it possibly involves a user-facing rename, we need to decide now.

psarna commented:
@avikivity let's use a neutral name, it doesn't hurt. I'll send v5 with it, but please also let me know your thoughts about the u32 vs u64 question that Vlad raised above.

avikivity commented:
Missed it, added my 2 bits

psarna commented:
v5:
 * removed the `_cql_` infix from the name, so that it can be used for other protocols (alternator, redis, etc.)
 * decided to leave uint32_t as is in order to limit potentially endless changes to this series

Also, I'll add support for this config variable to alternator (and maybe redis) in another series. I guess for a start it would be enough to treat this config value as independent per protocol (i.e. max number of requests per shard for a single protocol), assuming that nobody is currently planning to use Scylla for more than one protocol at the same time.

psarna commented:
I see that alternator can use DynamoDB's `RequestLimitExceeded` for the same purpose.

vladzcloudius commented on transport/server.hh:
```diff
@@ -104,6 +104,7 @@ struct cql_query_state {
 struct cql_server_config {
     ::timeout_config timeout_config;
     size_t max_request_size;
+    uint32_t max_concurrent_requests;
```
Here too - using the uint32_t saves us nothing since the next value is going to be 64-bit aligned anyway.

vladzcloudius commented on transport/server.cc:
```diff
@@ -618,6 +618,11 @@ future<> cql_server::connection::process_request() {
                     f.length, mem_estimate, _server._max_request_size));
         }
 
+        if (_server._requests_serving > _server._config.max_concurrent_requests) {
```
I have second thoughts about it.
I think it's very dangerous to leave it like this, @psarna. There is no such thing as "reasonable value" - there is a robust and not robust code.
This one is not robust.
 
The proper way to do this trick is to define the `max_concurrent_requests` default value to be such that can never be achieved by `_requests_serving`. 

One way to do that is to define `_requests_serving` as u32 and `max_concurrent_requests` be u64 and then initialize `max_concurrent_requests` to limits<u64>::max().

This way this code would always work no matter what.

psarna replied:
`max_concurrent_requests` does not need to be `u64`. If `_requests_serving` is u32 and `_server._config.max_concurrent_requests` is equal to u32::max, then this condition is always false, so it never sheds any requests. Effectively we're in the same situation even now, because it's not feasible for any modern shard to handle over 4 billion concurrent requests at a time -- but, I already have a small pull request with amendments to this one (https://github.com/scylladb/scylla/pull/7299), so I'll just add a commit which makes `_requests_serving` a u32.

vladzcloudius replied:
According to https://en.cppreference.com/w/cpp/types/numeric_limits the limits<uint32>::max() is achievable by uint32 value so my point holds.

psarna replied:
Of course it's achievable -- but assuming that `limit` is equal to `u32::max` and `x` is u32, a condition `x > limit` is never true, even for `x == u32::max`. (I hope this message ends up properly formatted, typing from my phone)

vladzcloudius replied:
Oh, you are right! Missed that. My bad.
So, yes, if we make `_requests_serving` u32 we should be totally safe.

vladzcloudius replied:
Hmmm... @psarna Are you planning to send a followup patch that would make _requests_serving u32 to address my comments above?

It would be better if this patch is reverted and you make the change in `_requests_serving` first otherwise it may break the `git bisect` if you try to catch a bug we mentioned above in the future.

psarna replied:
A followup was already sent yesterday: https://github.com/scylladb/scylla/pull/7299
